ape_hardhat/providers.py

```diff
@@ -308,34 +310,69 @@ class HardhatMainnetForkProvider(HardhatProvider):
     to use as your archive node.
     """
 
-    def _create_process(self) -> HardhatProcess:
-        mainnet_fork = self.config.mainnet_fork or {}  # type: ignore
-        upstream_provider_name = mainnet_fork.get("upstream_provider")
-        fork_block_num = mainnet_fork.get("block_number")
-        mainnet = self.network.ecosystem.mainnet
+    @property
+    def _fork_config(self) -> ConfigItem:
+        return self.config.mainnet_fork or {}  # type: ignore
+
+    @property
+    def upstream_provider_name(self) -> Optional[str]:
+        return self._fork_config.get("upstream_provider")
+
+    @property
+    def fork_block_number(self) -> Optional[int]:
+        return self._fork_config.get("block_number")
+
+    @property
+    def _mainnet(self) -> NetworkAPI:
+        return self.network.ecosystem.mainnet
 
-        # NOTE: if `upstream_provider_name` is `None`, this gets the default mainnet provider.
-        upstream_provider = mainnet.get_provider(provider_name=upstream_provider_name)
+    @cached_property
+    def _upstream_provider(self) -> ProviderAPI:
+        # NOTE: if 'upstream_provider_name' is 'None', this gets the default mainnet provider.
+        upstream_provider = self._mainnet.get_provider(provider_name=self.upstream_provider_name)
+        return upstream_provider
 
-        if not isinstance(upstream_provider, UpstreamProvider):
+    def connect(self):
+        super().connect()
+
+        # Check to see what the latest block number should be.
+        self._upstream_provider.connect()
+        latest_block_number = self._upstream_provider.get_block("latest").number
+        self._upstream_provider.disconnect()
+
+        # Verify that we connected to a Hardhat node with mainnet-fork mode.
+        buffer_range = self._mainnet.required_confirmations
+        re_org_range = range(latest_block_number - buffer_range, latest_block_number + buffer_range)
+        if self.get_block("latest").number not in re_org_range:
```

Comment:
why not?
```suggestion
        if self._upstream_provider.network != self.network:
```

Not sure if two `NetworkAPI` objects can be compared though

Reply:
That will always be `False` I think. `self.network` will always be `mainnet-fork` ad the upstream provider will be like `mainnet`

Reply:
The issue happen when you have a hardhat node proces going (maybe you left an `ape console --network ::hardhat` going. Then, you start another process using `ape console --network :mainnet-fork:hardhat`. It notices you have hardhat running on the port you said you would, so the network check outs! But, what is wrong is that is not actually a mainnet-fork, you can tell because usually the block number is like 0 or really small.

Reply:
> That will always be `False` I think. `self.network` will always be `mainnet-fork` ad the upstream provider will be like `mainnet`

ahhhh, I gotcha now. why not compare chain id?
```suggestion
        if self._upstream_provider.network.chain_id != self.network.chain_id:
```

Reply:
It seems as though the chain ID for a forked mainnet fork is not mainnet's chain ID:

```python
(harambe) (base) ➜  ape-demo-project git:(main) ape console --network :mainnet-fork:hardhat
INFO: Started Hardhat node at port '8545'.

In [1]: networks.active_provider.chain_id
Out[1]: 31337

```

Reply:
bizarre, hmm

Reply:
I think that makes sense, right? Like this is not really mainnet, this is a new network.

Reply:
Yeah, it makes sense, if that's how they are representing it. You could probably compare the genesis block hash for some semblance of determining what chain you are on.

Reply:
that worked